wremnants/CardTool.py

```diff
@@ -597,6 +598,112 @@ def writeForProcesses(self, syst, processes, label, check_systs=True):
         if syst != self.nominalName:
             self.fillCardWithSyst(syst)
 
+    def setSimultaneousABCD(self):
+        # Having 1 process for fakes, for each bin 3 free floating parameters, 2 normalization for lowMT and highMT and one fakerate between iso and anti iso
+        logger.info(f"Set processes for simultaneous ABCD fit")
+
+        # expected fake contribution
+        hist_fake = sum([group.hists[self.nominalName] if name == "Data" else -1*group.hists[self.nominalName] for name, group in self.datagroups.groups.items()])
+        
+        # setting errors to 0
+        hist_fake.view(flow=True)[...] = np.stack((hist_fake.values(flow=True), np.zeros_like(hist_fake.values(flow=True))), axis=-1)
+
+        if "passIso" not in hist_fake.axes.name or "passMT" not in hist_fake.axes.name:
+            raise RuntimeError('"passIso" and "passMT" expected to be found in histogram, but only have axes ')
+
+        # axes in the correct ordering
+        axes = ["passIso", "passMT"]
+        axes = [*axes, "charge"] if "charge" not in self.project else axes
+        axes += [ax for ax in self.project if ax not in axes]
+
+        if set(hist_fake.axes.name) != set(axes) or hist_fake.axes.name[0] != "passIso" or hist_fake.axes.name[1] != "passMT":
+            logger.debug("Axes in histogram are not the same as required or in a different order than expected, try to project")
+            hist_fake = hist_fake.project(*axes)
+
+        # set the expected values in the signal region
+        hist_fake.view(flow=True)[1,1,...] = sel.fakeHistABCD(hist_fake).view(flow=True)
+        
+        fakename = "Nonprompt"
+
+        self.datagroups.addGroup(fakename, label = "Nonprompt", color = "grey", members=[],)
+        self.datagroups.groups[fakename].hists[f"{self.nominalName}"] = hist_fake
+
+        bin_sizes = [ax.size for ax in hist_fake.axes if ax.name in axes and ax.name not in ["passIso","passMT"]]
+        axes = [ax.name for ax in hist_fake.axes if ax.name in axes and ax.name not in ["passIso","passMT"]]
+
+        for i in range(np.product(bin_sizes)):
+            current_i = i
+            ax_idx = []
+            for num in reversed(bin_sizes):
```

Comment:
I don't really understand what this is doing

Reply:
What's the logic of the bin index code? I don't follow it

Reply:
It loops in the reversed order as the product for the reco bins to get the outer index first. Lets say we have axes pT and eta with 2 and 3 bins, the product is 6, if we are at bin number 5 it takes the modulo of 3 first, which is 2, the index for the eta, then divides by this to have the rest. which is 1, in the next iteration it takes 1 modulo 2 which is 1, the index for pT